driver/modbus/scan_task.cpp

```
@@ -0,0 +1,127 @@
+                         << ": " << err;
+        devices_out.push_back(dev);
+    }
+    return {devices_out, xerrors::NIL};
+}
+
+bool Scanner::exec(
+    task::Command &cmd,
+    const synnax::Task &,
+    const std::shared_ptr<task::Context> &
+) {
+    if (cmd.type == TEST_CONNECTION_CMD_TYPE) {
+        this->test_connection(cmd);
+        return true;
+    }
+    return false;
+}
+
+xerrors::Error Scanner::check_device_health(synnax::Device &dev) const {
+    const auto rack_key = synnax::rack_key_from_task_key(this->task.key);
+    const auto parser = xjson::Parser(dev.properties);
+    const auto conn_cfg = device::ConnectionConfig(parser.child("connection"));
+    if (parser.error()) {
+        dev.status = synnax::DeviceStatus{
+            .key = dev.status_key(),
+            .name = dev.name,
+            .variant = status::variant::WARNING,
+            .message = "Invalid device properties",
+            .description = parser.error().message(),
+            .time = telem::TimeStamp::now(),
+            .details = {.rack = rack_key, .device = dev.key},
+        };
+        return parser.error();
+    }
+
+    auto [conn, conn_err] = this->devices->acquire(conn_cfg);
+    if (conn_err)
+        dev.status = synnax::DeviceStatus{
+            .key = dev.status_key(),
+            .name = dev.name,
+            .variant = status::variant::WARNING,
+            .message = "Failed to reach device",
+            .description = conn_err.message(),
+            .time = telem::TimeStamp::now(),
+            .details = {.rack = rack_key, .device = dev.key},
+        };
+    else
+        dev.status = synnax::DeviceStatus{
+            .key = dev.status_key(),
+            .name = dev.name,
+            .variant = status::variant::SUCCESS,
+            .message = "Device connected",
+            .time = telem::TimeStamp::now(),
+            .details = {.rack = rack_key, .device = dev.key},
+        };
+    return xerrors::NIL;
```

`check_device_health()` always returns `xerrors::NIL` even when connection fails. This hides errors from the caller. While the device status is correctly set to WARNING, the function signature suggests it should return the actual error. Should probably return `conn_err`